File: LeetCode/Greedy/Rust/Leetcode768-max-chunks-to-make-sorted-ii.rs
use std::collections::HashMap;
impl Solution {
    pub fn max_chunks_to_sorted(arr: Vec<i32>) -> i32 {
        let (mut b, n, mut mp) = (arr.clone(), arr.len(), HashMap::new());
        b.sort();
        let (mut s, mut res) = (0, 0);
        for i in 0..n {
            let v = mp.entry(arr[i]).or_insert(0);
            if *v == 1 {
                s -= 1;
            } else if *v == 0 {
                s += 1;
            }
            *v -= 1;
            let v = mp.entry(b[i]).or_insert(0);
            if *v == -1 {
                s -= 1;
            } else if *v == 0 {
                s += 1;
            }
            *v += 1;
            if s == 0 {
                res += 1;
            }
        }
        res
    }
}

File: LeetCode/Greedy/Python/Leetcode768-max-chunks-to-make-sorted-ii.py
class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        b, n = arr[:], len(arr)
        b.sort()
        res, s, mp = 0, 0, defaultdict(int)
        for i in range(n):
            if mp[arr[i]] == 1: s -= 1
            elif mp[arr[i]] == 0: s += 1
            mp[arr[i]] -= 1
            if mp[b[i]] == -1: s -= 1
            elif mp[b[i]] == 0: s += 1
            mp[b[i]] += 1
            if not s: res += 1
        return res
